src/sa.py

```diff
@@ -7,8 +14,187 @@ def main():
+    '''
+    suf = [x[i:] for i, _ in enumerate(x)]
+    suf = sorted(suf)
+    return [len(x) - len(i) for i in suf]
+
+
+def lower_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for lower bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if x[(sa[mid] + i) % len(x)] < a:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def upper_bound(a: str, i: int, lo: int, hi: int, x: str, sa: list[int]) -> int:
+    ''' binary search suffix array for upper bound for occurence of character a at position i in a pattern, in string x, given Suffix array SA of x'''
+    while hi > lo:
+        mid = (hi+lo)//2
+        if ord(x[(sa[mid] + i) % len(x)]) < ord(a)+1:
+            lo = mid+1
+        else:
+            hi = mid
+    return lo
+
+
+def pattern_match(x: str, p: str, sa: list[int]) -> Iterable[int]:
+    ''' report all occurences of pattern p in string x, given Suffix array: SA of x. '''
+    high = len(sa)
+    low = 0
+
+    if len(p) == 0 or len(x) == 0:
+        return []
+    for i, a in enumerate(p):
+        low = lower_bound(a, i, low, high, x, sa)
+        high = upper_bound(a, i, low, high, x, sa)
+
+    return sa[low:high]
+
+
+# Work in progress below
+def radix_by_index(x: str, u: list[int], alphabet: list[str]) -> dict[str, str]:
+    buckets = {i: 0 for i in alphabet}
+    triplets = get_triplet_map(x, u)
+
+    arr = [i for i in u]  # make copy so we don't override u
+    sa = [0] * len(u)
+    for i in range(3):
+        for trip in arr:
+            buckets[triplets[trip][i]] += 1
+        accsum = 0
+        for bucket in buckets:
+            buckets[bucket], accsum = accsum, accsum + buckets[bucket]
+        for trip in arr:
+            sa[buckets[triplets[trip][i]]] = trip
+            buckets[triplets[trip][i]] += 1
+        for bucket in buckets:
+            buckets[bucket] = 0
+        arr, sa = sa, arr
+
+    i = 0
+    sigma: dict[str, str] = {}
+    for trip in arr:
+        if triplets[trip] not in sigma:
+            sigma[triplets[trip]] = str(i)
+            i += 1
+    sigma = {str(sigma[key]): key for key in sigma}
+    return sigma
+
+
+def get_triplet_map(x: str, u: list[int]) -> dict[int, str]:
+    triplets: dict[int, str] = {}
+    for suf in u:
+        suffix = x[suf:]
+        if len(suffix) >= 3:
+            triplet = suffix[:3]
+        else:
+            triplet = suffix + (3-len(suffix)) * suffix[-1]
+        triplets[suf] = triplet
+    return triplets
+
+
+def skew(x: str) -> list[int]:
+    sa_0: list[int] = []
+    sa_1: list[int] = []
+    sa_2: list[int] = []
+    sa_12: list[int] = []
+    for i, _ in enumerate(x):
+        if i % 3 == 0:
+            sa_0.append(i)
+        elif i % 3 == 1:
+            sa_1.append(i)
+        else:
+            sa_2.append(i)
```

Comment:
One could also do

```python
sa_0 = [i for i in range(0, len(x), 3)]
sa_1 = [i for i in range(1, len(x), 3)]
sa_2 = [i for i in range(2, len(x), 3)]
```

or

```python
sa_12 = [i for i in range(1,len(x)) if i % 3]
```

(you don't really need `sa_1` or `sa_2` though).